### ISSUE TITLE
Minimalist example projects with tests: D3 Choropleth

### ISSUE THREAD

**Issue by QuincyLarson** [FreeCodeCamp/CurriculumExpansion#66]
Note: We may want to replace this graph with a more practical advanced visualization that involves less boilerplate code. Any thoughts on this, and what a suitable replacement might be?

Soon, campers will be able to build all front end projects right on FreeCodeCamp.com (no need for CodePen). We will also use automated tests to check whether campers have completed all of the agile user stories (and code review will become optional). 

This means we need to redesign these projects to have user stories that can be tested. And we should err on the side of having lots of tests (I'm thinking ~10 tests per project).

At the same time, we've heard feedback that campers are intimidated by the highly polished example projects we use. Instead of having fancy example projects, we should have minimalistic project examples that focus on functionality rather than appearance, animations, etc.

So we need:
- [ ] New agile user stories with one automated test each (reply to this issue with the user stories and their tests)
- [ ] Minimal example project that fulfills all of these user stories (reply with a link to CodePen)

The current project: https://www.freecodecamp.com/challenges/map-data-across-the-globe

**Comment by Christian-Paul** [comment]
@paycoguy and I were discussing potential D3 projects to replace this one and we both think that a Choropleth could fit the bill. Perhaps something like this: http://bl.ocks.org/mbostock/4060606.

The code doesn't look overly complex, so there should be less boilerplate, but the project is advanced enough to replace the Globe project at the end of the Data Viz section (and they're both geographical visualizations). 

Choropleths are frequently displayed in the New York Times and I think it would be practical for campers to learn this.

What are your thoughts? @QuincyLarson @no-stack-dub-sack

**Comment by QuincyLarson** [comment]
@paycoguy Great - I see these types of charts frequently. I agree that this is a great choice for a D3 project. Do you have any idea how you might go about testing this?

**Comment by Christian-Paul** [comment]
@QuincyLarson Sorry about the late response; we had to become familiar with the project to figure out how to test it.

We're going to be using a combination of approaches from the heat map and the bar chart to make tests for this one.

We'll test for the presence of components such as the title, description, legend, and areas (counties). We'll also check the areas and legend to make sure they have an adequate number of different fill colors.   Since the map should always be structured in the same way, we'll know what order the areas should be in and we'll check that the area id and data value is consistent with the data we provide. Finally, we'll use the same tooltip tests from our previous projects to make sure the tooltips are displayed properly and contain the correct information.

**Comment by QuincyLarson** [comment]
@Christian-Paul awesome. This sounds like a good approach. I'm thrilled that you're able to re-use techniques you've used on prior projects.

**Comment by no-stack-dub-sack** [comment]
@Christian-Paul @paycoguy Have you guys had a chance to check out the repo I've created for these? We should get together for a chat at some point to discuss this and the points you guys brought up in #64 about re-using certain tests globally.

**Comment by Christian-Paul** [comment]
@no-stack-dub-sack We've cloned the repo and have been developing this project and #65 here:
https://github.com/paycoguy/testable-projects-fcc

I take it I should copy the code onto codepen so that campers can see the example/use the suite on codepen, right?

Right now, some tests rely on access to the data, which we stored locally; it's not available online because we had to heavily parse it. Is there somewhere we can store the data for production?

I'd love to discuss re-using tests; I think the majority of the D3 tests can be re-used

**Comment by Christian-Paul** [comment]
Another one down! @no-stack-dub-sack @QuincyLarson 

Here is the pen with the tests: 
http://codepen.io/Christian-Paul/pen/ygNPjZ

And the sample project: 
http://codepen.io/Christian-Paul/pen/XpbzEv

In the repo environment, our changes are here:
https://github.com/paycoguy/testable-projects-fcc/tree/feature/choropleth

On the codepen version, I'm requesting the data from my personal site; how should I handle this properly?

# User Stories

1. My Choropleth should have a title with a corresponding id=\"title\"
2. My Choropleth should have a description with a corresponding id=\"description\"
3. My Choropleth should have counties with a corresponding class=\"county\" that represent the data
4. There should be at least 4 different fill colors used for the counties
5. My counties should each have \"data-fips\" and \"data-education\" properties containing their corresponding fips and education values
6. My Choropleth should have a county for each provided data point
7. The counties should have data-fips and data-education values that match the sample data
8. My Choropleth should have a legend with a corresponding id=\"legend\"
9. There should be at least 4 different fill colors used for the legend
10. When I mouse over a county, a \"div\" element with a corresponding id=\"tooltip\" should become visible
11. My tooltip should have a \"data-education\" property that corresponds to the given education of the active county

**Comment by QuincyLarson** [comment]
@Christian-Paul Awesome. This looks great! So do the tests. 

I've created the forum post here: https://forum.freecodecamp.com/t/d3-choropleth-project-with-testable-user-stories-guinea-pigs-needed/73922

One of the tests isn't passing for me: ![](https://forum.freecodecamp.com/t/d3-choropleth-project-with-testable-user-stories-guinea-pigs-needed/73922)

If you'll send me the dataset, I can throw it onto S3 and give you the link.

**Comment by Christian-Paul** [comment]
@QuincyLarson Thanks!

TopoJSON Data:
http://christianpaul.xyz/counties.json

Education Data:
http://christianpaul.xyz/education.json

Test 10 triggers mouse events and checks for the visibility of the tooltip. I get that error if I run the test and then manually mouse over the chart, since the tooltip will actually be visible, so that's likely what's happening.

**Comment by no-stack-dub-sack** [comment]
@Christian-Paul @paycoguy  Hey guys, this is awesome! I love the new project choice. Sorry for my delayed responses lately, Ive been incredibly busy between work and trying to get some other contributions done to help us get the new curriculum into beta... but that's finally done so I can turn my attention back to this a bit more. I also think I would be an excellent candidate for QAing these since I have not yet built any of these projects myself, so hopefully I can get started on that soon.

As far as the new project goes - since it is ready, do you want to try filing a PR and seeing how the merge goes? I haven't done a whole lot of managing PRs on my end, so it will be good experience too. Let m know, and I'll keep my eye out for one. I looked over the code very briefly, and you guys have the right idea, not that different at all.

**Comment by no-stack-dub-sack** [comment]
@QuincyLarson Hey Quincy, I unlisted this forum discussion for the time being since we do not yet have a blank pen that has this test suite available and since we do not yet have links for the JSON files either on S3 or as github raw data (like for the already existing projects). 

Once @Christian-Paul and @paycoguy file a PR against the repo and the test suite is integrated, we can get this suite in the blank pen for testing and can edit your OP on the forum and re-list this.

**Comment by no-stack-dub-sack** [comment]
@QuincyLarson Also, I am working on adding copy for each project to the repo for when we actually integrate these projects into the curriculum. Since things are working a little differently with the new testable projects, I figured new copy explaining things will be needed. Here's an example of what I have in mind (this is all html, not markdown):

<h2>Intro:</h2>
<p>For this project, we will be utilizing D3 to make a Choropleth Map:</p>
<blockquote cite="https://en.wikipedia.org/wiki/Choropleth_map">
	A <strong>choropleth</strong> map (from Greek χῶρος ("area/region") + πλῆθος ("multitude")) is a thematic map in which areas are shaded or patterned in proportion to the measurement of the statistical variable being displayed on the map, such as population density or per-capita income.
</blockquote>
<p>We will be using the following datasets for this project:</p>
<ul>
	<li>...need link</li>
	<li>...need link</li>
</ul>
<h2>The rules:</h2>
<ul>
	<li>
		Following the user stories below, and using the data sets we have linked to above, create a data visualization using D3 that is similar to this example: http://codepen.io/Christian-Paul/full/XpbzEv.
	</li>
	<li>
		Pass every test! Each user story has a corresponding automated test. To access this test suite, you can build your project in a fork of this CodePen pen: <a target="_blank" href="http://codepen.io/FCC_test_suites/pen/JbgXab">http://codepen.io/FCC_test_suites/pen/JbgXab</a>, or, if you are developing your project elsewhere, you can pull in the test suite from this CDN in a <code>&lt;script&gt;</code> tag: <a target="_blank" href="https://gitcdn.link/repo/no-stack-dub-sack/testable-projects-fcc/master/build/bundle.js">https://gitcdn.link/repo/no-stack-dub-sack/testable-projects-fcc/master/build/bundle.js</a>. Click the menu icon in the top left corner of the page to pick the correct test suite.
	</li>
	<li>Do not look at the example project's code, recreate the project from scratch and give it your own personal style.</li>
	<li>Remember to use Read-Search-Ask if you get stuck.</li>
	<li>When you are finished, click the "I've completed this challenge" button and include a link to your project.</li>
	<li>You can get feedback on your project by sharing it with your fellow campers on the <a target="_blank" href="https://forum.freecodecamp.com/c/project-feedback">forum</a>.</li>
</ul>
<p>Good luck and happy coding!</p>
<h2>User Stories:</h2>
<ol>
	<li>My Choropleth should have a title with a corresponding <code>id="title"</code>.</li>
	<li>My Choropleth should have a description with a corresponding <code>id="description"</code>.</li>
	<li>My Choropleth should have counties with a corresponding <code>class="county"</code> that represent the data.</li>
	<li>There should be at least 4 different fill colors used for the counties.</li>
	<li>My counties should each have <code>"data-fips"</code> and <code>"data-education"</code> properties containing their corresponding fips and education values.</li>
	<li>My Choropleth should have a county for each provided data point.</li>
	<li>The counties should have <code>data-fips</code> and <code>data-education</code> values that match the sample data.</li>
	<li>My Choropleth should have a legend with a corresponding <code>id="legend"</code>.</li>
	<li>There should be at least 4 different fill colors used for the legend.</li>
	<li>When I mouse over a county, a <code>&lt;div&gt;</code> element with corresponding <code>id="tooltip"</code> should become visible.</li>
	<li>My tooltip should have a <code>"data-education"</code> property that corresponds to the given education of the active county.</li>
</ol>

What do you think? We could use a slightly modified version of this for each project.

**Comment by QuincyLarson** [comment]
@no-stack-dub-sack agreed on de-listing the forum post for now.

Yes - I think it'd be good to have a description on the GitHub repo. At the same time, we want a single source of truth so we won't have to worry about conflicting or out-of-date information. I plan to add all of this information to the project description on FreeCodeCamp.com, so maybe instead we could just link to that page in the documentation?

**Comment by no-stack-dub-sack** [comment]
@QuincyLarson That's totally fine, I guess my main point was just to get started on the copy until there was an official place to put it. Are you ok with the way this looks or is this something you prefer to handle yourself? I'm ok with it either way, just thought I'd get a head start. We can discuss in greater detail offline soon.

**Comment by QuincyLarson** [comment]
@no-stack-dub-sack actually, now that I read your copy, I think it's great. I think we can use similar copy on the back end challenges, too. I'll talk with @Em-Ant and see what his thoughts are on this.

**Comment by no-stack-dub-sack** [comment]
@QuincyLarson Ok, great, sounds good. Thanks!

**Comment by no-stack-dub-sack** [comment]
@paycoguy @Christian-Paul That was easy! I merged the PR - thanks guys! I also pulled out the `getTooltipStatus` and `getRandomIndex` functions into the entry point and added them to the global library object so all the D3 tests can now share those functions - so that's a bit less repeated code, but I have yet to figure out how to share the whole tooltip test among suites. But it's a step in the right direction... I've forked the project over to this collection: http://codepen.io/collection/nMjJjE/ - this is where all the example projects live, all the D3 ones can be viewed there (I've credited you guys in the comments of the JS and we can add more thorough ones later). 

Also, I added a copy of the education.json file that's not being exported so we can just link to that raw github data for users to use. And I changed the link in the forked project so its no longer pulling from your site @Christian-Paul. 

@QuincyLarson I've re-listed this forum discussion now that the test suite is integrated and we have viable links to the data sets!

Thanks for your hard work guys - awesome job on these, seriously, they look freaking great!

**Comment by QuincyLarson** [comment]
@no-stack-dub-sack Awesome - thank you for filling us in on this, and for re-listing the forum post.

By the way, @Christian-Paul @paycoguy if these forum posts aren't getting much attention yet, don't worry - the will. Once we get the curriculum expansion live on beta.freecodecamp.com, I'll make a big announcement to the entire community to stimulate interest in QA'ing them.

**Comment by QuincyLarson** [comment]
@no-stack-dub-sack here is the copy we agreed upon for the front end projects:

```
[ Challenge title ]
[ Screenshot ]

Fulfill the user stories by getting all of the tests to pass. Use whichever libraries you need. Give it your own personal style.

You can build your project in a fork [this CodePen pen](http://codepen.io/FCC_test_suites/pen/JbgXab) or use this CDN to run the tests in any environment you like: <code>https://gitcdn.link/repo/no-stack-dub-sack/testable-projects-fcc/master/build/bundle.js</code>

Here's a working example. Try not to look at its code: http://codepen.io/Christian-Paul/full/XpbzEv

Datasets: [ dataset URLs if necessary]

Remember to use Read-Search-Ask if you get stuck.
```

**Comment by no-stack-dub-sack** [comment]
@QuincyLarson The pen link that is in this boiler plate is not the official one I forked over to FCC, I missed this when we were going over all of this. I just saw your PR [#12570](https://github.com/freeCodeCamp/freeCodeCamp/pull/12570) and noticed all the projects were included - this URL will have to be switched out at some point. One of us can submit a PR once this one is merged into staging. Can't wait to check everything out!

**Comment by QuincyLarson** [comment]
@no-stack-dub-sack OK - I can quickly change those out. What is the URL for the official boilerplate pen?

**Comment by no-stack-dub-sack** [comment]
@QuincyLarson Sorry, didn't see this til now and looks like its already been merged. I just opened an issue though and can fix it if you'd like. Also, we're not doing this line anymore?

> ...or use this CDN to run the tests in any environment you like: <code>https://gitcdn.link/repo/no-stack-dub-sack/testable-projects-fcc/master/build/bundle.js</code>
